### DISCUSSION TITLE
How can I get the specific time（Accurate to the millisecond） of  photos on the report exported from the O3d312？How can I translate the "TimestampNsec" and "TimestampSec"  to this format（for example : 2024-08-28 21:17:35）?

### DISCUSSION THREAD

**Discussion by mikufans-dumo** [ifm/ifm3d#464] [Q&A] [answered]
**Here‘s the report.**
[O3D3xx_2024-07-18_094602_00-02-01-44-d7-f9.zip](https://github.com/user-attachments/files/16775646/O3D3xx_2024-07-18_094602_00-02-01-44-d7-f9.zip)
**For example**，how can I get the specific time（Accurate to the millisecond） of the first fail photo？ 
![Snipaste_2024-08-28_10-30-15](https://github.com/user-attachments/assets/c96f5f57-1ccf-44d7-acba-3a3465f7766e)

**How can I translate the   "TimestampNsec" and "TimestampSec"  to this format（for example : 2024-08-28 21:17:35）?**

**Here's the data of first fail photo。**

{
    "ApplicationDecodingResult": 0,
    "ApplicationID": "476707713",
    "ApplicationIndex": 8,
    "ApplicationName": "Completeness application1",
    "DeviceName": "New sensor",
    "ExposureTime1": 200,
    "ExposureTime2": 1000,
    "LogiclayerExecutionStatus": 1,
    "MotionCompensatedImage": 0,
    "RtspOverlayStyle": "{\"ROI\": {\"default\": {\"visible\": true, \"filled\": false, \"use_symbol\": false, \"label_alignment\": \"top\", \"label_content\": \"\", \"label_background\": \"black\", \"label_fontsize\": 8, \"label_failonly\": false}, \"model_defaults\": {}, \"specific\": {} } }",
    "TemplateInfo": "{\n    \"ROI_cols\": 6,\n    \"ROI_exposure\": true,\n    \"ROI_grid\": 1,\n    \"ROI_roi\": true,\n    \"ROI_rows\": 4,\n    \"ROI_shape\": 3,\n    \"boundaryOverfill\": 0.013754600000000001,\n    \"detectOverfill\": true,\n    \"detectUnderfill\": true,\n    \"goodStateTeached\": true,\n    \"movingSpeed\": false,\n    \"negateOverfillOut\": false,\n    \"negateUnderfillOut\": false,\n    \"optimizedWithROIs\": true,\n    \"overfillStateTeached\": true,\n    \"pulsDuration1\": 20,\n    \"pulsDuration2\": 20,\n    \"signalMode1\": false,\n    \"signalMode2\": false,\n    \"type\": \"Completeness\",\n    \"underfillStateTeached\": true,\n    \"version\": \"1.0.0\",\n    \"wizard_completed\": true\n}\n",
    "TimestampNsec": 580195647,
    "TimestampSec": 946684897,
    "Type": "WithModels",
    "adv_statistics": {
        "average_decoding_time_of_all_found_models_ms": 0,
        "average_decoding_time_of_not_all_found_models_ms": 0,
        "frame_acquisition": {
            "binsize": 2,
            "max": 42,
            "mean": 27.27380952381,
            "min": 16,
            "standard_deviation": 4.332466553813,
            "variance": 18.77026643991
        },
        "frame_duration": {
            "binsize": 2,
            "max": 393,
            "mean": 351.9880952381,
            "min": 346,
            "standard_deviation": 6.985519116305,
            "variance": 48.79747732426
        },
        "frame_evaluation": {
            "binsize": 2,
            "max": 393,
            "mean": 351.9880952381,
            "min": 346,
            "standard_deviation": 6.985519116305,
            "variance": 48.79747732426
        },
        "maximum_decoding_time_of_all_found_models_ms": 0,
        "maximum_decoding_time_of_not_all_found_models_ms": 0,
        "minimum_decoding_time_of_all_found_models_ms": 0,
        "minimum_decoding_time_of_not_all_found_models_ms": 0,
        "model_statistics": {
            "model_001": {
                "numGood": 85,
                "numInvalid": 0,
                "numOverSP2": 1,
                "numUnderSP1": 6
            }
        },
        "number_of_aborted_frames": 0,
        "number_of_all_found_models": 0,
        "number_of_failed_frames": 7,
        "number_of_frames": 85,
        "number_of_not_all_found_models": 0,
        "number_of_passed_frames": 78,
        "number_of_timeouts": 0,
        "version": 3
    },
    "ll_results": {
        "digital_out": {
            "mask": 7,
            "values": 1
        },
        "statistics": {
            "current_passed": false,
            "num_failed": 27,
            "num_passed": 93
        }
    },
    "model_001": {
        "Index": 1,
        "Name": "Completeness",
        "ROIs": {
            "0": {
                "ROIName": "Complet/0",
                "procval": -0.04473799467087,
                "quality": 1,
                "state": 0
            },
            "1": {
                "ROIName": "Complet/1",
                "procval": -0.05417466163635,
                "quality": 1,
                "state": 0
            },
            "2": {
                "ROIName": "Complet/2",
                "procval": -0.04773080348969,
                "quality": 1,
                "state": 0
            },
            "3": {
                "ROIName": "Complet/3",
                "procval": -0.04728710651398,
                "quality": 1,
                "state": 0
            },
            "4": {
                "ROIName": "Complet/4",
                "procval": -0.0476753115654,
                "quality": 1,
                "state": 0
            },
            "5": {
                "ROIName": "Complet/5",
                "procval": -0.03574329614639,
                "quality": 1,
                "state": 0
            },
            "6": {
                "ROIName": "Complet/6",
                "procval": -0.03369987010956,
                "quality": 1,
                "state": 0
            },
            "7": {
                "ROIName": "Complet/7",
                "procval": -0.0336035490036,
                "quality": 1,
                "state": 0
            },
            "8": {
                "ROIName": "Complet/8",
                "procval": -0.02391654253006,
                "quality": 0.9361702203751,
                "state": 0
            },
            "9": {
                "ROIName": "Complet/9",
                "procval": -0.03198766708374,
                "quality": 0.9285714030266,
                "state": 0
            },
            "10": {
                "ROIName": "Complet/10",
                "procval": -0.02706104516983,
                "quality": 1,
                "state": 0
            },
            "11": {
                "ROIName": "Complet/11",
                "procval": -0.02273160219193,
                "quality": 1,
                "state": 0
            },
            "12": {
                "ROIName": "Complet/12",
                "procval": -0.1118894219398,
                "quality": 1,
                "state": 7
            },
            "13": {
                "ROIName": "Complet/13",
                "procval": -0.03926193714142,
                "quality": 0.9424460530281,
                "state": 0
            },
            "14": {
                "ROIName": "Complet/14",
                "procval": -0.02685928344727,
                "quality": 0.905797123909,
                "state": 0
            },
            "15": {
                "ROIName": "Complet/15",
                "procval": -0.02060294151306,
                "quality": 0.8848921060562,
                "state": 0
            },
            "16": {
                "ROIName": "Complet/16",
                "procval": -0.02440971136093,
                "quality": 0.9214285612106,
                "state": 0
            },
            "17": {
                "ROIName": "Complet/17",
                "procval": -0.01228177547455,
                "quality": 1,
                "state": 0
            },
            "18": {
                "ROIName": "Complet/18",
                "procval": -0.02856355905533,
                "quality": 1,
                "state": 0
            },
            "19": {
                "ROIName": "Complet/19",
                "procval": -0.02457481622696,
                "quality": 0.9568345546722,
                "state": 0
            },
            "20": {
                "ROIName": "Complet/20",
                "procval": -0.01351374387741,
                "quality": 0.9219858050346,
                "state": 0
            },
            "21": {
                "ROIName": "Complet/21",
                "procval": -0.01482039690018,
                "quality": 0.892086327076,
                "state": 0
            },
            "22": {
                "ROIName": "Complet/22",
                "procval": -0.0122703909874,
                "quality": 0.9347826242447,
                "state": 0
            },
            "23": {
                "ROIName": "Complet/23",
                "procval": -0.01114702224731,
                "quality": 0.955882370472,
                "state": 0
            }
        },
        "ROIsUpdated": {
            "ROIs": [
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 42.33894729614,
                        "xMin": 28.64938545227,
                        "yMax": 44.47579193115,
                        "yMin": 31.34735870361
                    },
                    "center_x": 35.49416732788,
                    "center_y": 37.91157531738,
                    "group": 0,
                    "height": 12.99778556824,
                    "id": 0,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 60.90256118774,
                        "xMin": 47.21300125122,
                        "yMax": 44.29615020752,
                        "yMin": 31.16771697998
                    },
                    "center_x": 54.05778121948,
                    "center_y": 37.73193359375,
                    "group": 1,
                    "height": 12.99778556824,
                    "id": 1,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 79.46617889404,
                        "xMin": 65.77660369873,
                        "yMax": 44.11650848389,
                        "yMin": 30.98807525635
                    },
                    "center_x": 72.62139129639,
                    "center_y": 37.55229187012,
                    "group": 2,
                    "height": 12.99778556824,
                    "id": 2,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 98.02979278564,
                        "xMin": 84.34021759033,
                        "yMax": 43.93687438965,
                        "yMin": 30.80844116211
                    },
                    "center_x": 91.18500518799,
                    "center_y": 37.37265777588,
                    "group": 3,
                    "height": 12.99778556824,
                    "id": 3,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 116.5934066772,
                        "xMin": 102.9038467407,
                        "yMax": 43.75723266602,
                        "yMin": 30.62879943848
                    },
                    "center_x": 109.748626709,
                    "center_y": 37.19301605225,
                    "group": 4,
                    "height": 12.99778556824,
                    "id": 4,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 135.1570129395,
                        "xMin": 121.4674606323,
                        "yMax": 43.57759094238,
                        "yMin": 30.44915771484
                    },
                    "center_x": 128.3122406006,
                    "center_y": 37.01337432861,
                    "group": 5,
                    "height": 12.99778556824,
                    "id": 5,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 42.51310348511,
                        "xMin": 28.82354164124,
                        "yMax": 62.4727897644,
                        "yMin": 49.34435653687
                    },
                    "center_x": 35.66832351685,
                    "center_y": 55.90857315063,
                    "group": 6,
                    "height": 12.99778556824,
                    "id": 6,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 61.07671737671,
                        "xMin": 47.38715744019,
                        "yMax": 62.29315185547,
                        "yMin": 49.16471862793
                    },
                    "center_x": 54.23193740845,
                    "center_y": 55.7289352417,
                    "group": 7,
                    "height": 12.99778556824,
                    "id": 7,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 79.64032745361,
                        "xMin": 65.95076751709,
                        "yMax": 62.11351013184,
                        "yMin": 48.9850769043
                    },
                    "center_x": 72.79554748535,
                    "center_y": 55.54929351807,
                    "group": 8,
                    "height": 12.99778556824,
                    "id": 8,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 98.20394134521,
                        "xMin": 84.51438140869,
                        "yMax": 61.9338684082,
                        "yMin": 48.80543518066
                    },
                    "center_x": 91.35916137695,
                    "center_y": 55.36965179443,
                    "group": 9,
                    "height": 12.99778556824,
                    "id": 9,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 116.7675552368,
                        "xMin": 103.0779953003,
                        "yMax": 61.75423049927,
                        "yMin": 48.62579727173
                    },
                    "center_x": 109.9227752686,
                    "center_y": 55.1900138855,
                    "group": 10,
                    "height": 12.99778556824,
                    "id": 10,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 135.3311767578,
                        "xMin": 121.6416244507,
                        "yMax": 61.57459259033,
                        "yMin": 48.44615936279
                    },
                    "center_x": 128.4864044189,
                    "center_y": 55.01037597656,
                    "group": 11,
                    "height": 12.99778556824,
                    "id": 11,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 42.68725967407,
                        "xMin": 28.9976978302,
                        "yMax": 80.46979522705,
                        "yMin": 67.34136199951
                    },
                    "center_x": 35.84247970581,
                    "center_y": 73.90557861328,
                    "group": 12,
                    "height": 12.99778556824,
                    "id": 12,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 61.25087738037,
                        "xMin": 47.56131744385,
                        "yMax": 80.29014587402,
                        "yMin": 67.16171264648
                    },
                    "center_x": 54.40609741211,
                    "center_y": 73.72592926025,
                    "group": 13,
                    "height": 12.99778556824,
                    "id": 13,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 79.81449127197,
                        "xMin": 66.12493133545,
                        "yMax": 80.11051177979,
                        "yMin": 66.98207855225
                    },
                    "center_x": 72.96971130371,
                    "center_y": 73.54629516602,
                    "group": 14,
                    "height": 12.99778556824,
                    "id": 14,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 98.37810516357,
                        "xMin": 84.68854522705,
                        "yMax": 79.93087005615,
                        "yMin": 66.80243682861
                    },
                    "center_x": 91.53332519531,
                    "center_y": 73.36665344238,
                    "group": 15,
                    "height": 12.99778556824,
                    "id": 15,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 116.9417190552,
                        "xMin": 103.2521591187,
                        "yMax": 79.75123596191,
                        "yMin": 66.62280273438
                    },
                    "center_x": 110.0969390869,
                    "center_y": 73.18701934814,
                    "group": 16,
                    "height": 12.99778556824,
                    "id": 16,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 135.5053253174,
                        "xMin": 121.8157730103,
                        "yMax": 79.57159423828,
                        "yMin": 66.44316101074
                    },
                    "center_x": 128.6605529785,
                    "center_y": 73.00737762451,
                    "group": 17,
                    "height": 12.99778556824,
                    "id": 17,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 42.86141967773,
                        "xMin": 29.17185783386,
                        "yMax": 98.46678161621,
                        "yMin": 85.33834838867
                    },
                    "center_x": 36.01663970947,
                    "center_y": 91.90256500244,
                    "group": 18,
                    "height": 12.99778556824,
                    "id": 18,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 61.42502975464,
                        "xMin": 47.73546981812,
                        "yMax": 98.28714752197,
                        "yMin": 85.15871429443
                    },
                    "center_x": 54.58024978638,
                    "center_y": 91.7229309082,
                    "group": 19,
                    "height": 12.99778556824,
                    "id": 19,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 79.98863983154,
                        "xMin": 66.29907989502,
                        "yMax": 98.10750579834,
                        "yMin": 84.9790725708
                    },
                    "center_x": 73.14385986328,
                    "center_y": 91.54328918457,
                    "group": 20,
                    "height": 12.99778556824,
                    "id": 20,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 98.55226898193,
                        "xMin": 84.86269378662,
                        "yMax": 97.9278717041,
                        "yMin": 84.79943847656
                    },
                    "center_x": 91.70748138428,
                    "center_y": 91.36365509033,
                    "group": 21,
                    "height": 12.99778556824,
                    "id": 21,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 117.1158828735,
                        "xMin": 103.4263076782,
                        "yMax": 97.74822998047,
                        "yMin": 84.61979675293
                    },
                    "center_x": 110.2710952759,
                    "center_y": 91.1840133667,
                    "group": 22,
                    "height": 12.99778556824,
                    "id": 22,
                    "type": "Ellipse",
                    "width": 13.56442451477
                },
                {
                    "anchor_bound": 1,
                    "angle": -0.5544325993763,
                    "boundingBox": {
                        "xMax": 135.679473877,
                        "xMin": 121.9899215698,
                        "yMax": 97.56858825684,
                        "yMin": 84.4401550293
                    },
                    "center_x": 128.8347015381,
                    "center_y": 91.00437164307,
                    "group": 23,
                    "height": 12.99778556824,
                    "id": 23,
                    "type": "Ellipse",
                    "width": 13.56442451477
                }
            ]
        },
        "SP1": -0.057145,
        "SP2": 0.022425,
        "Type": "Level",
        "anchorFound": true,
        "hasAnchorTracking": true,
        "numGood": 23,
        "numInvalid": 0,
        "numOverSP2": 0,
        "numUnderSP1": 1
    }
}

**Comment by debertst** [comment]
Hi @mikufans-dumo 
The system time can be converted with the following code:

   timeval tv;
   // "timestampSec" aus dem Chunk-Header
   tv.tv_sec=0x70DC6559;
   tv.tv_usec=1;

   time_t nowtime=tv.tv_sec;
   struct tm *nowtm;
   char buf[64];

   nowtm = localtime(&nowtime);
   strftime(buf, sizeof buf, "%Y-%m-%d %H:%M:%S", nowtm);
   printf("%s\n", buf);

**Reply by mikufans-dumo** [reply]
Thank you for your reply @debertst ，But I still have problem about the conversion time。
I first translate the 946684897（"TimestampSec"）to 0x386d43e1. Then I use your code to get the time：2000-01-01 08:01:37。
****The problem** is** this time is not the report exported time。You can see the report name is O3D3xx_2024-07-18_094602_00-02-01-44-d7-f9，this report is exported on 2024-07-18. And I promise the recorded photos and exported report are created on the same day.
![Snipaste_2024-09-02_16-11-04](https://github.com/user-attachments/assets/e5258f00-8663-4297-a8b2-157de8f8aacd)
![Snipaste_2024-09-02_16-13-24](https://github.com/user-attachments/assets/d713c78a-05db-4c1a-b5f1-33337960a185)

**Comment by debertst** [comment, marked as answer]
Hi @mikufans-dumo, 

for a correct time, you have to activate the NTP on the O3D
![image](https://github.com/user-attachments/assets/67455471-a21a-41dc-9c21-b65111e240d4)


Without this, a default time of the sensor is displayed, e.g. 2021-09-02 with the current firmware 1.71.10121. Which firmware are you using?
I would recommend FW 1.71.10121
